### PR TITLE
Fix plugin java class desugar

### LINKED ISSUE #72185: [google_maps_flutter] fatal exception and crash on Android: java.lang.AbstractMethodError
After upgrading to Flutter (Channel dev, 1.25.0-8.0.pre), when I start google maps (google_maps_flutter: 1.0.6), I have an issuer and crash:

```bash
java.lang.AbstractMethodError: abstract method "void io.flutter.plugin.platform.PlatformView.onFlutterViewAttached(android.view.View)"
E/AndroidRuntime(21906): 	at io.flutter.plugin.platform.VirtualDisplayController.onFlutterViewAttached(VirtualDisplayController.java:181)
E/AndroidRuntime(21906): 	at io.flutter.plugin.platform.PlatformViewsController$1.createVirtualDisplayForPlatformView(PlatformViewsController.java:233)
E/AndroidRuntime(21906): 	at io.flutter.embedding.engine.systemchannels.PlatformViewsChannel$1.create(PlatformViewsChannel.java:104)
E/AndroidRuntime(21906): 	at io.flutter.embedding.engine.systemchannels.PlatformViewsChannel$1.onMethodCall(PlatformViewsChannel.java:59)
E/AndroidRuntime(21906): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/AndroidRuntime(21906): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/AndroidRuntime(21906): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:738)
E/AndroidRuntime(21906): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/AndroidRuntime(21906): 	at android.os.MessageQueue.next(MessageQueue.java:326)
E/AndroidRuntime(21906): 	at android.os.Looper.loop(Looper.java:160)
E/AndroidRuntime(21906): 	at android.app.ActivityThread.main(ActivityThread.java:6669)
E/AndroidRuntime(21906): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(21906): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/AndroidRuntime(21906): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
I/Process (21906): Sending signal. PID: 21906 SIG: 9
Lost connection to device.
```

flutter doctor -v
```bash
[✓] Flutter (Channel dev, 1.25.0-8.0.pre, on macOS 11.0.1 20B29 darwin-x64, locale ru-RU)
    • Flutter version 1.25.0-8.0.pre at 
    • Framework revision a12e2a473a (3 days ago), 2020-12-08 23:13:04 -0500
    • Engine revision df39e5c515
    • Dart version 2.12.0 (build 2.12.0-133.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at 
    • Platform android-30, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Connected device (1 available)
    • AOSP on IA Emulator (mobile) • emulator-5554 • android-x86 • Android 9 (API 28) (emulator)

• No issues found!

    dependencies {
        classpath 'com.android.tools.build:gradle:4.1.1'
        classpath 'com.google.gms:google-services:4.3.4'
    }
```
Note: 
Tested 
Android 9, Android 10, Android 11 - exception and crash
Android 6 - ok

### PR DESCRIPTION
**Pull request by blasten**
Fixes https://github.com/flutter/flutter/issues/72185.

I think this change will cause an issue if flavors are used. Unfortunately, if that's the case, we will have no choice, but to remove the support.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: dev/prod_builders.json [hunk 1]
```diff
@@ -264,6 +264,12 @@
          "task_name": "linux_fuchsia_precache",
          "flaky": false
       },
+      {
+        "name": "Linux gradle_desugar_classes_test",
+        "repo": "flutter",
+        "task_name": "linux_gradle_desugar_classes_test",
```
Comment (blasten) [inline dev/prod_builders.json:270]: @godofredoc I was going over https://github.com/flutter/flutter/blob/master/dev/README.md, is there anything else required to configure the prod/try builders?

File: packages/flutter_tools/gradle/flutter.gradle [hunk 2]
```diff
@@ -318,25 +318,16 @@ class FlutterPlugin implements Plugin<Project> {
...
             }
-            // The embedding is a compileOnly dependency of a Flutter plugin,
-            // however prior to Gradle 6.0.0, it must be an API dependency.
+            // The embedding is API dependency of the plugin, so the AGP is able to desugar
+            // default method implementations when the interface is implemented by a plugin.
             //
-            // Not doing so, causes transitive dependency resolution conflicts.
-            // That is, the embedding dependencies resolved in the plugin are
-            // different than the ones resolved in the app.
-            if (isGradleVersionGraterOrEqualThan('6.0.0')) {
```
Comment (dnfield) [inline packages/flutter_tools/gradle/flutter.gradle:327]: nit: just noticed this but that method should be `Greater`, not `Grater`.

Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle:327]: ahh. I removed this method as it's no longer used.

**Comment by dnfield** [comment]
You are right that this has broken flavors.

I'm fairly certain that there are people using this. I'm not sure I grasp the full scope of the problem here - I suspect it's some confluence of flavors being something hacky that was never fully baked/thought through/owned and gradle madness.

/cc @xster for thoughts about killing support for flavors on Android. It might also be worth consulting with the Gradle/AGP team to see if there's some way we can still make this worth.

**Comment by xster** [comment]
I don't think we added specific analytics but based on github issues, flavors and build configs on iOS are definitely used. Can you describe the root issue a bit and how this solution was chosen?

**Comment by blasten** [comment]
I left a comment in the Gradle script. Basically, after https://issuetracker.google.com/139821726, AGP has a new optimization that configures the build such as Java 8+ APIs are desugared if an explicit API dependency is set on such code.   Here the API is the Android embedding, which is used by a plugin. 

When two Gradle projects have the same API dependency, you get duplicated symbols https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8858351162389675568/+/steps/run_gradle_plugin_light_apk_test/0/stdout?format=raw

I will ask the AGP team to confirm that my understanding is correct.

**Comment by blasten** [comment]
Alright. The good news is that I was able to fix the issue following a suggestion from the AGP team.   

The new solution is: If an app has plugin, the app doesn't depend on the embedding directly. Instead, the app sets an API dependency on the plugins, and the plugins set an API dependency on the embedding. Let me know if that makes sense.

ptal

**Comment by dnfield** [comment]
Will that work out ok if there are multiple plugins?

**Comment by blasten** [comment]
Yes, it does

**Comment by blasten** [comment]
Yes, it will

File: packages/flutter_tools/gradle/flutter.gradle [hunk 3]
```diff
@@ -199,21 +211,13 @@ class FlutterPlugin implements Plugin<Project> {
...
-            : "$hostedRepository/download.flutter.io"
-        project.rootProject.allprojects {
-            repositories {
-                maven {
-                    url repository
-                }
-            }
+        // The embedding is set as an API dependency in a Flutter plugin.
+        // Therefore, don't make the app project depend on the embedding if there are Flutter
+        // plugins since it causes duplicated classes.
```
Comment (dnfield) [inline packages/flutter_tools/gradle/flutter.gradle:216]: nit: maybe mention the interplay with the flavors implementation here.

**Review by dnfield — APPROVED** [review]
LGTM

File: dev/prod_builders.json (hunk 1 above)
Comment (godofredoc) [reply dev/prod_builders.json:270]: Yes if this is a new tasks then the builders should exist before adding these configurations. My recommendation is to land the code changes then create the configs in infra and the final step would be updating the prod|try_builders.json.

We just added some docs: https://g3doc.corp.google.com/company/teams/flutter/luci/builders.md?cl=head

Comment (blasten) [reply dev/prod_builders.json:270]: cool

**Pull request merged**